pulsar-broker/src/main/java/org/apache/pulsar/broker/service/StickyKeyConsumerSelector.java

```diff
@@ -33,13 +33,15 @@ public interface StickyKeyConsumerSelector {
      *
      * @param consumer new consumer
      */
-    CompletableFuture<Void> addConsumer(Consumer consumer);
+    CompletableFuture<Map<Consumer, NavigableSet<Range>>> addConsumer(Consumer consumer);
```

Comment:
Good point! Thanks for asking.

I avoided initially introducing types since the Pulsar code base isn't very type rich. We tend to write very procedural code and continue to do so. However, I think it would be useful to use types to convey concepts that are abstractions. Using collection types with a lot of generics is usually a code smell.

I have refactored the code to use types. In this case `Map<Consumer, NavigableSet<Range>>` is replaced with `ImpactedConsumersResult` which "Represents the consumers that were impacted by a hash range change in a StickyKeyConsumerSelector at a point in time". That can `processRemovedHashRanges` which processes each impacted `Consumer` and the associated `RemovedHashRanges` which "Represents the hash ranges which were removed from an existing consumer by a change in the hash range assignments.".

During this refactoring, I also made a `ConsumerHashAssignmentsSnapshot` type which "Represents the hash ranges assigned to each consumer in a StickyKeyConsumerSelector at a point in time." In a StickyKeyConsumerSelector implementation, the hash ranges are snapshotted each time a consumer is added or removed. The ConsumerHashAssignmentsSnapshot contains a method `resolveImpactedConsumers` which returns the `ImpactedConsumersResult`.

Do the refactorings make it more understandable?

Reply:
Refactored and discussed with @merlimat